packages/@aws-cdk/aws-sagemaker/lib/endpoint.ts

```diff
@@ -0,0 +1,448 @@
+      resourceName: 'AWSServiceRoleForApplicationAutoScaling_SageMakerEndpoint',
+    }));
+  }
+}
+
+abstract class EndpointBase extends cdk.Resource implements IEndpoint {
+  /**
+   * The ARN of the endpoint.
+   *
+   * @attribute
+   */
+  public abstract readonly endpointArn: string;
+
+  /**
+   * The name of the endpoint.
+   *
+   * @attribute
+   */
+  public abstract readonly endpointName: string;
+
+  /**
+   * Permits an IAM principal to invoke this endpoint
+   * @param grantee The principal to grant access to
+   */
+  public grantInvoke(grantee: iam.IGrantable) {
+    return iam.Grant.addToPrincipal({
+      grantee,
+      actions: ['sagemaker:InvokeEndpoint'],
+      resourceArns: [this.endpointArn],
+    });
+  }
+}
+
+/**
+ * Construction properties for a SageMaker Endpoint.
+ */
+export interface EndpointProps {
+
+  /**
+   * Name of the endpoint.
+   *
+   * @default - AWS CloudFormation generates a unique physical ID and uses that ID for the
+   * endpoint's name.
+   */
+  readonly endpointName?: string;
+
+  /**
+   * The endpoint configuration to use for this endpoint.
+   */
+  readonly endpointConfig: IEndpointConfig;
+}
+
+/**
+ * HTTP response codes for Endpoint invocations
+ */
+export enum InvocationHttpResponseCode {
+  /**
+   * 4xx response codes from Endpoint invocations
+   */
+  INVOCATION_4XX_ERRORS = 'Invocation4XXErrors',
+
+  /**
+   * 5xx response codes from Endpoint invocations
+   */
+  INVOCATION_5XX_ERRORS = 'Invocation5XXErrors',
+}
+
+/**
+ * Defines a SageMaker endpoint.
+ */
+export class Endpoint extends EndpointBase {
+  /**
+   * Imports an Endpoint defined either outside the CDK or in a different CDK stack.
+   * @param scope the Construct scope.
+   * @param id the resource id.
+   * @param endpointArn the ARN of the endpoint.
+   */
+  public static fromEndpointArn(scope: Construct, id: string, endpointArn: string): IEndpoint {
+    const endpointName = cdk.Stack.of(scope).splitArn(endpointArn, cdk.ArnFormat.SLASH_RESOURCE_NAME).resourceName!;
+
+    class Import extends EndpointBase {
+      public endpointArn = endpointArn;
+      public endpointName = endpointName;
+    }
+
+    return new Import(scope, id, {
+      environmentFromArn: endpointArn,
+    });
+  }
+
+  /**
+   * Imports an Endpoint defined either outside the CDK or in a different CDK stack.
+   * @param scope the Construct scope.
+   * @param id the resource id.
+   * @param endpointName the name of the endpoint.
+   */
+  public static fromEndpointName(scope: Construct, id: string, endpointName: string): IEndpoint {
+    const endpointArn = cdk.Stack.of(scope).formatArn({
+      service: 'sagemaker',
+      resource: 'endpoint',
+      resourceName: endpointName,
+    });
+
+    class Import extends EndpointBase {
+      public endpointArn = endpointArn;
+      public endpointName = endpointName;
+    }
+
+    return new Import(scope, id, {
+      environmentFromArn: endpointArn,
+    });
```

Comment:
I don't like this duplication. We should have a third method, `fromEndpointAttributes` where the `Import` class lives, and both `fromEndpointArn` and `fromEndpointName` call that method.

Reply:
Should I mark this `fromEndpointAttributes` method as private? I ask because [the following design guideline](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#from-attributes) implies that the method should only be exposed for non-name, non-ARN scenarios:

> If a resource has more than a single attribute (“ARN” and “name” are usually considered a single attribute since it's usually possible to convert one to the other), then the resource should provide a static fromAttributes method to allow users to explicitly supply values to all resource attributes when they import an external (unowned) resource [awslint:from-attributes].

Reply:
I'm reading that design guideline as `fromXxxAttributes` should be added if a resource has more than a single attribute, but not specifically disallowing the method when thats not the case. The reason for this linter rule is likely to guard against the duplication we have here when we have multiple `fromXxxThing` methods.

I am for adding in a public `fromXxAttributes`, unless our linter rules somehow flag that as wrong, which would surprise me.